cognite/extractorutils/_extensions.py

```diff
@@ -0,0 +1,187 @@
+#  Copyright 2022 Cognite AS
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""
+A module containing utilities for developing extensions for generic source systems.
+"""
+
+from datetime import datetime
+from types import TracebackType
+from typing import Callable, Iterable, List, Optional, Tuple, Type, Union
+
+from more_itertools import peekable
+
+from cognite.client import CogniteClient
+from cognite.client.data_classes import Event as _Event
+from cognite.client.data_classes import Row as _Row
+from cognite.extractorutils.base import Extractor
+from cognite.extractorutils.configtools import CustomConfigClass
+from cognite.extractorutils.metrics import BaseMetrics
+from cognite.extractorutils.statestore import AbstractStateStore
+from cognite.extractorutils.uploader import EventUploadQueue, RawUploadQueue, TimeSeriesUploadQueue
+
+try:
+    from typing import TypeAlias  # type: ignore
+except ImportError:
+    # Backport for python < 3.10
+    from typing_extensions import TypeAlias
+
+
+class RawRow:
+    def __init__(self, db_name: str, table_name: str, row: Union[_Row, Iterable[_Row]]):
+        self.db_name = db_name
+        self.table_name = table_name
+        if isinstance(row, Iterable):
+            self.rows = row
+        else:
+            self.rows = [row]
+
+
+TimeStamp = Union[int, datetime]
+
+
+class InsertDatapoints:
+    def __init__(
+        self,
+        *,
+        id: Optional[int] = None,
+        external_id: Optional[str] = None,
+        datapoints: Union[List[Tuple[TimeStamp, float]], List[Tuple[TimeStamp, str]]],
+    ):
+        self.id = id
+        self.external_id = external_id
+        self.datapoints = datapoints
+
+
+Event: TypeAlias = _Event
+
+CdfTypes = Union[Event, Iterable[Event], RawRow, Iterable[RawRow], InsertDatapoints, Iterable[InsertDatapoints]]
```

Comment:
These are a bit tricky. They will be needed by users of extensions, both to instantiate instances and for typing, so having them in a private module will yield all sorts of linting errors on every usage 🤔

Reply:
We could have every extension export them through a typing module just containing type aliases for these, but then it's yet another list to keep in sync with this